lib/puppet/provider/service/daemontools.rb

```diff
@@ -53,7 +53,6 @@ def defpath
             break
           end
         end
-        raise "Could not find the daemon directory (tested [/var/lib/service,/etc])" unless @defpath
```

Comment:
Ideally we could confine the provider based on the existence on one of the daemon directories, so that the framework never calls the unsuitable provider. Something like:

```
confine :exists => defpath
```

And then change `defpath` to:

```ruby
def self.defpath
  ["/var/lib/service", "/etc"].find do |path|
    Puppet::FileSystem.exist?(path) && FileTest.directory?(path)
  end
end
```

If we can't confine based on those directories (see the [init provider's comment](https://github.com/puppetlabs/puppet/blob/e2aa511e5e8f2a48440e09c52330da6001f35df4/lib/puppet/provider/service/init.rb#L30)), then I think we should check for `defpath` returning nil, otherwise we'll call `FileTest.directory?(nil)` which happens to be nil-safe. Suggest checking for nil explicitly:

```ruby
def self.instances
  path = self.defpath
  unless path
    Puppet.info("daemontools is unsuitable because service directory is nil")
  end
  ...
end
```

Reply:
There are no other confines on `defpath` so i opt on using the second options as i do not have enough context on why the confine on defpath is problematic. Having the `attr_writer :defpath` means that the defpath can be set from different places.